x/incentive/types/adapter.go

```diff
@@ -0,0 +1,17 @@
+package types
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
+// SourceAdapter provides source shares from an external module.
+type SourceAdapter interface {
+	// GetShares returns source shares owned by one address.
+	//
+	// For example, the shares a user owns in the kava:usdx and bnb:usdx swap pools.
+	// It returns the shares for several sources at once, in the same order as the sourceIDs. Specifying no sourceIDS will return no shares.
+	GetShares(ctx sdk.Context, owner sdk.AccAddress, sourceIDs []string) []sdk.Dec
```

Comment:
Thoughts on making this return a map instead? Returning a slice that must be the same order seems a bit more easier to make errors compared to dealing with iterating over a map deterministically.

e.g. in the spike using it, each adapter implementation must be careful to return the exact order vs. if returning a map, we may be able to localize the deterministic iteration to 1 method.

https://github.com/Kava-Labs/kava/blob/49e8e4b1c4378858765ae70ddbf33896c5a89bd4/x/incentive/keeper/rewards_swap.go#L115-L117

Reply:
Yeah we could try it out. I don't have a strong opinon, both aren't great.
Good point that new adapters will be written more frequently than code consuming the interface.
We'll have to be careful to check each value exists in the map when reading from it, otherwise it'll return `sdk.Dec{}` which panics whenever it's used.
If we're only using it one method a map is probably ok. We've had several indeterminsm bugs from map iteration, and I think Juno went down partially due to it as well. Seems to be really easy to forget about.